Bluejay.asm

```diff
@@ -3539,18 +3541,38 @@ pgm_start:
 	call	wait250ms
 	sjmp	startup_beep_done
 
+; Startup tune has 50 pairs of (item1, item2) - a total of 100 items.
+; item2 - is the duration of each pulse of the musical note, lower the value, higher the pitch
+; item1 - if item2 is zero, it is the number of milliseconds of wait time, else it is the number of pulses of item2
 startup_beep_melody:
-	call	beep_f1					; Normal startup melody
-	call	wait5ms
-	call	beep_f2
-	call	wait5ms
-	call	beep_f1
-	call	wait5ms
-	call	beep_f3
-	call	wait200ms
-	call	beep_f2
-	call	beep_f4
-	call	beep_f4
+	mov	Temp1,	#32h
+	mov	DPTR,	#Eep_Pgm_Startup_Tune
+
+startup_melody_loop:
+	; Read current location at Eep_Pgm_Startup_Tune to Temp4. If Temp4 is 0, no point trying to play this note
+	clr	A
+	movc	A,	@A+DPTR
+	mov	Temp4,	A				; TODO: See if flash eeprom value can be read directly into Temp4
+	jz	startup_melody_loop_next_note	; TODO: Have to see why these having zero notes is bricking the ESC
+	inc	DPTR
+
+	; Read location at Eep_Pgm_Startup_Tune Temp1 to Temp3. If temp3 is zero, that means this is a silent note
+	clr	A
+	movc	A,	@A+DPTR
+	mov	Temp3,	A				; TODO: See if flash eeprom can be read directly into Temp3
+	jz	startup_melody_item_wait_ms	; TODO: Have to see why these having zero notes is bricking the ESC
+	inc	DPTR
+	call	beep
+	sjmp	startup_melody_loop_next_note
+
+startup_melody_item_wait_ms:
+	mov	A,	Temp4
+	mov	Temp2,	A
+	call	wait_ms_o
```

Comment:
The call to `wait_ms_o` mutates `Temp1` so it should not be used as counter.
Notice how the general pattern is that registers are not pushed to the stack, so be careful with jumps and calls.

Reply:
Good catch. Pushed the fixes. The ESC is now able to play the current start up tune properly. Still couldn't rescue my bricked one. Probably my bad soldering skills are to blame, will look at it later. For now i have 2 more ESCs to brick.

The next todo I have already implemented but haven't pushed is:

* Remove the additional setting to select between startup beep vs. startup tune as we can use the startup tune for everything. Should i push this change? That will cause a change in Eeprom layout. Or do you want to reuse that variable space for something else?

The main todo i am still figuring out how to is:

Make the `Eep_Pgm_Startup_Tune` configurable by the configurator.

* Why are eeprom read/write functions split into block 1/block 2? Is there any limit to the amount of data that can be read/written at once?
* I keep getting `*** WARNING L30: MEMORY SPACE OVERLAP` when i moved the startup tune array before the `Eep_Dummy` (even after adjusting the next CSEG's location.) What am i missing?

Can you share any tips on how to implement this task?

Reply:
- I'm planning to change the eeprom layout but have left it mostly backward compatible with BLHeli for now since there are a number of assumptions about it in the code and the configurator. We probably need to decide if we want to merge this beep melody feature without too many eeprom changes, or if we should wait merging it until we have a new eeprom layout structure.
- The read/write blocks are probably legacy code from blheli. I think maybe block 2 was only used on some escs? Also some mcus with real eeprom might have had different limitations.
- Maybe related to reserved space in `Common.inc` at `1A40h` and `1A50h`? This is part of what I'm planning to change as well.

Reply:
>  We probably need to decide if we want to merge this beep melody feature without too many eeprom changes, or if we should wait merging it until we have a new eeprom layout structure.

I don't mind it either way, but some bugs may pop up only after you integrate the "make this configurable with configurator" feature, (like total byte count of parameters + tune becoming > 255 and then having to split things up). So reserving specific address for the melody data seems to make more sense. 

For the code part, it seems like making a change for a very specific use case and would be nice to have something more "generic/extensible" like "ESC variable parameters" or some such thing, which just reserves some writable buffer that can be filled with values. Then we can directly reserve like 250 bytes for everything someone may want to add in the future.

Reply:
@mathiasvr Any updates on the eeprom layout changes? Will that come on top of the layout related refactoring?

Reply:
The eeprom layout changes are actually a separate thing. The esc layout refactoring is a bit overkill but was something I was working on while trying to support a new type of esc pinout.

I'm not sure when to change eeprom layout because I would like to have settings mostly stable at that point (and do it only once). I hope you're not dependant on that? I'm fine if you reserve an address for the melody in this pr.

Reply:
Ah will try to implement read out of melody data via. configurator in this PR itself then. I think this will be a busy week for me, so most probably I will look into it during the weekend.

Reply:
Ah sorry, I forgot you wanted to look into the configurator already. I will look into the eeprom layout soon, I think many of the changes can be done and still be compatible. I can try to prepare the read/write of melody data as well?

Reply:
Hi, just a heads up. This weekend i looked at your forked blheli configurator code and started adding a Melody component to it. I think the interesting thing is i don't think we have to add any more changes to the ESC code to make it send the melody data specifically.

What's happening is:

`readbackSettings = (await _4way.read(BLHELI_SILABS_EEPROM_OFFSET, BLHELI_LAYOUT_SIZE)).params;`

So just updating BLHELI_LAYOUT_SIZE should work i think.

Yet to test it out, and my knowledge of react is also non existent at this point but this is turning out to be an interesting challenge so far. Learning a lot of things I wouldn't have usually touched.

Let me know if i need to be careful about more things. I think next week onwards my schedule should be more free to focus on these things better.

Reply:
I have been working a bit on this as well. Looks like we can only write 256 bytes at a time, so might be easiest to start with melodies within that range.

Reply:
But 256 bytes is plenty right?

![image](https://user-images.githubusercontent.com/265528/109796379-18c03b80-7c3e-11eb-969c-5ca78559e3cd.png)

I have some local changes i am yet to push (increasing the melody length from 100 bytes to 128 bytes - so we get 64 notes), early end the loop if the current note duration is 0. Will push them once i test this out.

Finally managed to get the configurator built - it was missing some OS X dependency (i am on linux). After moving it to optional dependency it built fine.